test/realtime/alerts_fetcher_test.exs

```diff
@@ -0,0 +1,218 @@
+          "service_effect" => "Non-detour alert"
+        },
+        "type" => "alert",
+        "id" => 789
+      }
+    ]
+  }
+
+  describe "start_link/1" do
+    test "starts GenServer" do
+      assert {:ok, _pid} = AlertsFetcher.start_link(update_fn: fn _ -> :ok end)
+    end
+  end
+
+  describe "init/1" do
+    test "sets state and queues up initial check" do
+      api_url = "https://test.com/"
+      api_key = "test_key"
+      reassign_env(:skate, :api_url, api_url)
+      reassign_env(:skate, :api_key, api_key)
+
+      update_fn = fn _ -> :ok end
+
+      assert {:ok,
+              %{
+                update_fn: ^update_fn,
+                poll_interval_ms: 500,
+                api_url: ^api_url,
+                api_key: ^api_key
+              },
+              {:continue, :initial_poll}} =
+               AlertsFetcher.init(update_fn: update_fn, poll_interval_ms: 500)
+    end
+  end
+
+  describe "handle_continue/2" do
+    test "fetches alerts" do
+      bypass = Bypass.open()
+      api_url = "http://localhost:#{bypass.port}/"
+      reassign_env(:skate, :api_url, api_url)
+      test_pid = self()
+
+      update_fn = fn alerts ->
+        GenServer.cast(test_pid, alerts)
+        :ok
+      end
+
+      {:ok, state, _} = AlertsFetcher.init(update_fn: update_fn)
+
+      Bypass.expect(bypass, fn conn ->
+        Plug.Conn.resp(conn, 200, Jason.encode!(@api_response))
+      end)
+
+      assert {:noreply, _state} = AlertsFetcher.handle_continue(:initial_poll, state)
+
+      assert_receive {:"$gen_cast", %{"1" => ["Route 1 detour"]}}, 5000
+    end
+  end
+
+  describe "handle_info/2" do
+    test "fetches alerts and logs on success" do
+      set_log_level(:info)
+      bypass = Bypass.open()
+      api_url = "http://localhost:#{bypass.port}/"
+      reassign_env(:skate, :api_url, api_url)
+      test_pid = self()
+
+      update_fn = fn alerts ->
+        GenServer.cast(test_pid, alerts)
+        :ok
+      end
+
+      {:ok, state, _} = AlertsFetcher.init(update_fn: update_fn)
+
+      Bypass.expect(bypass, fn conn ->
+        Plug.Conn.resp(conn, 200, Jason.encode!(@api_response))
+      end)
+
+      log =
+        capture_log([level: :info], fn ->
+          assert {:noreply, _state} = AlertsFetcher.handle_info(:query_api, state)
+        end)
+
+      assert_receive {:"$gen_cast", %{"1" => ["Route 1 detour"]}}, 5000
+
+      assert log =~ "updated_alerts"
+    end
+
+    test "handles unsuccessful HTTP request" do
+      bypass = Bypass.open()
+      api_url = "http://localhost:#{bypass.port}/"
+      reassign_env(:skate, :api_url, api_url)
+
+      update_fn = fn _ -> :ok end
+
+      {:ok, state, _} = AlertsFetcher.init(update_fn: update_fn)
+
+      Bypass.expect(bypass, fn conn ->
+        Plug.Conn.resp(conn, 500, "server error")
+      end)
+
+      log =
+        capture_log([level: :warn], fn ->
+          assert {:noreply, _state} = AlertsFetcher.handle_info(:query_api, state)
+        end)
```

Comment:
thought(non-blocking): I remember coming across [with_log](https://hexdocs.pm/ex_unit/1.14.2/ExUnit.CaptureLog.html#with_log/2) but have never actually used it. This could be a case for it so that the assertions can be kept together at the end of the test.

Reply:
Oh neat, I hadn't seen that before, will look into using it.